### PR TITLE
PAC (Proxy auto-config) support using pacparser library

### PR DESCRIPTION
**Pull request by jschwartzenberg**
This PR will probably require some more clean-up, but I'm creating it already to see what Codacy/AppVeyor says and to create awareness.

As with the other PR, I did not develop this, but I hope to get this merged! See here for my discussion with the original developer:
https://github.com/olfryd/pacntlm/issues/1

Information about the library can be found here: http://pacparser.manugarg.com/
RHEL builds of the library are here: https://copr.fedorainfracloud.org/coprs/jschwart/pacparser/

### CODE REVIEW

**Comment by jschwartzenberg** [comment]
@versat the build fails now on the missing pacparser library. Shall I add a compile time option to enable PAC support with a default of it being disabled?

**Comment by versat** [comment]
As far as I see there is no prepuilt pacparser package for cygwin. So I think the best way is to add a compile time option and disable it by default as you suggested. Maybe it can be enabled by default later when there is an easy way to get it built in cygwin and via AppVeyor or so.

**Review by jschwartzenberg — COMMENTED** [review]
The permission changes on `debian/rules` don't seem right. I'll revert that. Btw. I noticed you do squashed merged. I can rebase issues if that's preferred.

File: forward.c [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include <strings.h>
 #include <assert.h>
 
+#include "direct.h"
```
Comment (jschwartzenberg) [inline forward.c:32]: You think I should put `ifdef`s here?

Comment (versat) [reply forward.c:32]: No, i think that is fine.

**Comment by versat** [comment]
> Btw. I noticed you do squashed merged. I can rebase issues if that's preferred.

I have no strong opinion on this. IMO doing squashed merges of PRs is more clear when viewing the history. If you want a specific PR being merged in another way that should be no problem 🙂

**Comment by jschwartzenberg** [comment]
Another advantage is that `bisect` will work properly :) Doing a squashed merge is fine. I would otherwise rebase this all into a single commit, which would effectively be the same.

Let me know if you still see any things to fix before you merge this! Thanks!

**Comment by versat** [comment]
I managed to get the Travis build results beeing added to the checks of the PRs again. The permission changes let `make rpm` fail in Travis.

**Comment by jschwartzenberg** [comment]
Whoops yep, I had missed the `rpm/rules` one. I had only noticed the `debian/rules` one. This should be corrected now. It seems it was accidentally changed in the initial patch.

**Comment by versat** [comment]
Do you plan to work on the TODOs? If you don't mind i would add you as a contributor so you can make changes yourself, merge to Coverity branch to get results there and so on.

**Comment by jschwartzenberg** [comment]
Good question. So far I mainly checked that the patch works apart from solving the issues as the original commits are not by me (neither the TODOs).

I do intend to do my best to make this comply with quality standards that are in place and at the same time I hope to have at least one pair of extra eyes before anything would be merged to a main branch. Originally I had the intention to at least introduce the IPv6 support that is also going around in another repo.

I had originally not intended to do more than that, also not these TODOs, but as I plan to use this in a corporate environment I do have an incentive to sure things are in as good a shape as possible. Let me know if there are TODOs that you consider minimally important to solve before merging, I will look into addressing those for sure. I would also be glad to be a contributor if that you think it would help the project.

**Comment by versat** [comment]
Maybe the travis build can be extended without too many changes to also build cntlm with pacman. If you have time to add this that would be great, otherwise i will add it when i have time.

**Comment by jschwartzenberg** [comment]
I haven't got much experience with `pacman` yet. I started using Parabola at home a few months ago. It would be neat to get buids in a PPA, COPR, etc. and possibly in distro repos. I hope to look into that. I also noticed that many of the SonarQube issues should not be too hard to fix.

**Comment by olfryd** [comment]
For the pacparser library, I already created a package in COPR (https://copr.fedorainfracloud.org/coprs/nebu/pacntlm/). However, versat's cntlm with PAC support is still missing. Let me know if you would like to have the build scripts for pacparser to create your own COPR repo.

**Comment by jschwartzenberg** [comment]
@olfryd that would be very useful yes! The COPR repo I created for pacparser was just based on a single source RPM :) It will be good to have automated builds in place for RHEL/Fedora.

**Comment by olfryd** [comment]
You can find everything here: https://github.com/olfryd/pacparser. The pacparser.spec is under rpm/.

**Comment by jschwartzenberg** [comment]
> @jschwartzenberg You can find everything here: https://github.com/olfryd/pacparser. The pacparser.spec is under rpm/.

Thanks! I hope to find a moment to take a look soon! Btw. you may want to check your author info in those commits. (You can open a commit in GitHub and append `.patch` to the URL.)

**Comment by olfryd** [comment]
Thanks a lot for changing the mail address! And sorry for any inconvenience this may have caused.

> Thanks! I hope to find a moment to take a look soon! Btw. you may want to check your author info in those commits. (You can open a commit in GitHub and append .patch to the URL.)

Done. :) Thanks for pointing out.

**Comment by jschwartzenberg** [comment]
No problem! Thanks a lot for providing such a valuable patch! Hopefully you can contribute more in the future :)

**Pull request merged**